File: addons/guide-cpp/src/ui/virtual_joy/guide_virtual_joy_base.cpp
#include "guide_virtual_joy_base.h"
#include "../../guide.h"

using namespace godot;

void GUIDEVirtualJoyBase::_bind_methods() {
    BIND_ENUM_CONSTANT(MODE_TOUCH);
    BIND_ENUM_CONSTANT(MODE_MOUSE);
    BIND_ENUM_CONSTANT(MODE_MOUSE_AND_TOUCH);

    ADD_SIGNAL(MethodInfo("changed"));
    ADD_SIGNAL(MethodInfo("configuration_changed"));
}

GUIDEVirtualJoyBase::GUIDEVirtualJoyBase() {
}

GUIDEVirtualJoyBase::~GUIDEVirtualJoyBase() {
}

void GUIDEVirtualJoyBase::_input(const Ref<InputEvent> &event) {
    Ref<InputEventMouse> mouse_event = event;
    if (mouse_event.is_valid()) {
        if (input_mode == MODE_MOUSE || input_mode == MODE_MOUSE_AND_TOUCH) {
            _handle_mouse_input(mouse_event);
        }
    }

    if (event->is_class("InputEventScreenTouch") || event->is_class("InputEventScreenDrag")) {
        if (input_mode == MODE_TOUCH || input_mode == MODE_MOUSE_AND_TOUCH) {
            _handle_touch_input(event);
        }
    }
}

void GUIDEVirtualJoyBase::_handle_mouse_input(const Ref<InputEventMouse> &event) {
}

void GUIDEVirtualJoyBase::_handle_touch_input(const Ref<InputEvent> &event) {
}

Vector2 GUIDEVirtualJoyBase::_screen_to_world(Vector2 input) const {
    return get_canvas_transform().affine_inverse() * input;
}

void GUIDEVirtualJoyBase::_reconnect() {
    if (!is_node_ready()) return;
    if (_virtual_joy_id != 0) {
        GUIDE::get_singleton()->get_input_state()->disconnect_virtual_stick(_virtual_joy_id);
    }
    _virtual_joy_id = GUIDE::get_singleton()->get_input_state()->connect_virtual_stick(virtual_device);
}

File: addons/GuideCS/UI/VirtualJoy/GUIDEVirtualJoyBase.cs
using System;
using Godot;

namespace Guide;

public partial class GUIDEVirtualJoyBase : CenterContainer {
	[Signal] public delegate void ChangedEventHandler();
	[Signal] public delegate void ConfigurationChangedEventHandler();

	public enum InputMode {
		Touch,
		Mouse,
		MouseAndTouch
	}

	private int _virtualDevice = 0;
	[Export(PropertyHint.Enum, "Any:-1,1:0,2:1,3:2,4:3,Virtual 1:-2,Virtual 2:-3,Virtual 3:-4,Virtual 4:-5")]
	public int VirtualDevice {
		get => _virtualDevice;
		set {
			_virtualDevice = value;
			_Reconnect();
			EmitSignal(SignalName.ConfigurationChanged);
		}
	}

	private InputMode _inputMode = InputMode.Touch;
	[Export]
	public InputMode InputModeValue {
		get => _inputMode;
		set {
			_inputMode = value;
			EmitSignal(SignalName.ConfigurationChanged);
		}
	}

	private bool _drawDebug = false;
	[Export]
	public bool DrawDebug {
		get => _drawDebug;
		set {
			_drawDebug = value;
			QueueRedraw();
			EmitSignal(SignalName.ConfigurationChanged);
		}
	}

	internal bool _isActuated = false;
	protected int _virtualJoyId = 0;

	protected void _Reconnect() {
		if (!IsNodeReady()) return;

		if (_virtualJoyId != 0) {
			GUIDE.Singleton?._InputState.DisconnectVirtualStick(_virtualJoyId);
		}

		_virtualJoyId = GUIDE.Singleton?._InputState.ConnectVirtualStick(VirtualDevice) ?? 0;
	}

	public override void _Input(InputEvent @event) {
		if (@event is InputEventMouse mouseEvent) {
			if (InputModeValue == InputMode.Mouse || InputModeValue == InputMode.MouseAndTouch) {
				_HandleMouseInput(mouseEvent);
			}
		}

		if (@event is InputEventScreenTouch || @event is InputEventScreenDrag) {
			if (InputModeValue == InputMode.Touch || InputModeValue == InputMode.MouseAndTouch) {
				_HandleTouchInput(@event);
			}
		}
	}

	protected virtual void _HandleMouseInput(InputEventMouse @event) { }
	protected virtual void _HandleTouchInput(InputEvent @event) { }

	protected Vector2 _ScreenToWorld(Vector2 input) {
		return GetCanvasTransform().AffineInverse() * input;
	}
}
